[RFC] Calculate atom groups on GPU

HanatoK opened the pull request:
Hi @giacomofiorin and @jhenin ! This WIP PR is a partial implementation of the plan discussed in https://github.com/Colvars/colvars/discussions/816. This PR only implements the calculation in `cvm::atom_group` in CUDA when the proxy sets the flag `support_gpu` to true. The test code is in [tests/functional_gpu/run_colvars_test_cuda.cpp](https://github.com/Colvars/colvars/compare/master...HanatoK:colvars_gpu?expand=1#diff-6ff575eee2834fb3ad0e9e130a1398f8feed5bc78bcd8672b8d3ececcd4a0e9b). Needless to say, there are many still questions and issues:

**colvarproxy**
- ~~What's the best practice to allow a proxy to declare GPU support? Currently this PR uses the `support_gpu` flag in `colvarproxy_gpu` for it but sometimes people may prefer CPU even if the proxy supports GPU. Should we also check `smp`, or provide a way to set the `support_gpu` flag from the config file?~~ (Done, `smp gpu` is supported)

**colvarmodule** and **colvar**
- The logic of executing GPU kernels in this PR is running all kernels in a single stream but with possibly different graphs. To improve the parallelization, this PR unpacks all atom groups from CVCs in `colvar {...}` and computes them in a single CUDA graph, and then copies the data to the host memory if necessary. The CVCs and `colvar {...}` blocks are computed after that. In other words, the order of computation looks like reverse level order traversal of a tree, which is different from the CPU code. However, the current implementation relies partially on C++ exceptions, and I think it would be better to convert `colvardeps*` into an AST (as disscussed in https://github.com/Colvars/colvars/pull/700), and move the feature-level dependency resolution into a separate class/functor that traverses the AST.

**colvaratoms**
- I still think `colvaratoms` is too heavy and the optimal alignment should be implemented as a separate CVC, and Colvars should allow referencing it in other CVCs;
- I would like to incorporate the reading of total forces into the `read_data` graph, but `cvm::atom_group::read_total_forces` can be called anywhere from the CVC class, so it is difficult to know if the total forces are required beforehand. Is it possible to know the parents of an atom group requesting total forces in `read_data` of the atom group?

**tests**
- ~~How can I add [tests/functional_gpu/run_colvars_test_cuda.cpp](https://github.com/Colvars/colvars/compare/master...HanatoK:colvars_gpu?expand=1#diff-6ff575eee2834fb3ad0e9e130a1398f8feed5bc78bcd8672b8d3ececcd4a0e9b) to the cmake tests?~~
- ~~How can I compare the results of [tests/functional_gpu/run_colvars_test_cuda.cpp](https://github.com/Colvars/colvars/compare/master...HanatoK:colvars_gpu?expand=1#diff-6ff575eee2834fb3ad0e9e130a1398f8feed5bc78bcd8672b8d3ececcd4a0e9b) with [run_colvars_test.cpp](https://github.com/Colvars/colvars/blob/master/tests/functional/run_colvars_test.cpp)?~~
- The tests can be run by the following commands
```bash
cd tests/functional_gpu
mkdir build
cd build/
cmake ../
make -j8
make test
```

HanatoK committed:
feat/gpu: wrappers for GPU memory functions and stream APIs

HanatoK committed:
refactor: separate the colvarproxy_gpu and GPU helper functions

HanatoK committed:
build: fix the build files for NAMD and VMD

HanatoK committed:
feat: add cudaMemset wrappers and 32-bit GPU proxy atom buffers

HanatoK committed:
refactor: simplify the GPU management functions

HanatoK committed:
refactor: move the allocation and deallocation of GPU buffers to
colvaratoms_gpu.*

HanatoK committed:
Multiple updates (see the description below)

1. Implement do_feature_side_effects_gpu to allocate buffers for fit
   gradients
2. Add kernels for fitting group forces ("fit gradients" of non-scalar
   components)
3. Simplify the HIP defines
4. Try to support debug gradients

HanatoK committed:
fix: adapt the macros to fix the compilation

HanatoK committed:
fix: fix the compilation

HanatoK committed:
Multiple fixes and new debug code

1. Fix the kernel parameters in atoms_calc_cog;
2. Fix the initialization of rot_gpu;
3. Fix the apply_forces without fit gradients.
4. Output the CUDA graphs in case of cvm::debug().

HanatoK committed:
fix: fix the calc_fit_gradients node

HanatoK committed:
fix: fix the COG kernel

HanatoK committed:
test: add CUDA test

HanatoK committed:
build: fix the VMD build

HanatoK committed:
build: fix the build files for NAMD and VMD again

HanatoK committed:
build: add missing colvar_rotation_derivative.C

HanatoK committed:
build: fix the GROMACS compilation

HanatoK committed:
fix: fix a crash in the desctruction of rotation_gpu

HanatoK committed:
fix: fix the memory leaks in init_gpu

The init() function of an atom group can be called multiple times in
Colvars, so it is better to use reallocate_* to avoid the memory leaks.

HanatoK committed:
opt: merge the get_q kernel into the Jacobi kernel

HanatoK committed:
prof: add NVTX profiling support

HanatoK committed:
fix: fix the memory issue

1. Correctly use placement new for rot_deriv_gpu;
2. Avoid dangling pointer for h_sum_applied_colvar_force.

HanatoK committed:
Add experimental support of run Colvars on GPU in the CudaGM interface

HanatoK committed:
fix: some kernels are not necessary in case of Colvars GPU

HanatoK committed:
opt: skip the synchronization in case of Colvars GPU

HanatoK committed:
opt: skip the unnecessary sync in apply_forces

Since we use the default stream from the MD engine, the stream will be
synchronized by the MD engine anyway.

HanatoK committed:
Manually revert f5dfc95791d525cd382630fb53d35208e5187871 for the time
being

The code requires to copy the lattice immediately. This will be fixed
later.

HanatoK committed:
Clear the new colvar forces array and avoid the unnecessary CUDA sync

HanatoK committed:
build: update the cmake build file with the experimental Colvars GPU
flag

jhenin commented:
Thanks @HanatoK ! About user control, I think a behavior similar to the current smp flag would be good, that is, enabled by default if the hardware supports it, and can be disabled explicitly. If the value is a string, it opens the possibility for finer control, as you recently did with smp to allow for selecting the parallelization scheme.

jhenin commented:
I see your point about dependency resolution, and I agree that moving to an explicit AST is better.

HanatoK committed:
Reset the thread block atomic counters in the kernel

HanatoK committed:
opt: limit the number of blocks for reduction operations

In my benchmarks, it looks like limiting the number of blocks can
significantly improve the performance of reduction operations.

HanatoK committed:
Guard the new global variables

HanatoK committed:
fix: fix the reduction in calc_fit_forces_impl_loop1_kernel

HanatoK committed:
refactor: make the GPU support an SMP mode

This commit also refactors the SMP selection code. The function
get_available_smp_modes() is supposed to return the SMP modes that the
proxy supports, and get_preferred_smp_mode() should return the SMP mode
that the proxy prefers (the default SMP mode).

HanatoK committed:
build: fix the VMD compilation

HanatoK committed:
build: fix the VMD compilation again

HanatoK committed:
build: remove the unused CMake flag and definition

HanatoK commented:
> Thanks @HanatoK ! About user control, I think a behavior similar to the current smp flag would be good, that is, enabled by default if the hardware supports it, and can be disabled explicitly. If the value is a string, it opens the possibility for finer control, as you recently did with smp to allow for selecting the parallelization scheme.

Done in https://github.com/Colvars/colvars/pull/826/commits/7f36ea28a1ef4062e9cc5f063797e68ff0653b57.

HanatoK commented:
> At this point, there is only a moderate amount of code dedicated to the optimal alignment in the colvaratoms class. And the orientation CVC is already mostly equivalent to what you describe. So maybe this goal could be achieved by adapting the orientation CVC?

I am not sure. The orientation CVC only gives a quaternion, but the `cvm::atom_group` calculates a rotated frame. If we want to adapt the orientation CVC, then we may have to move a lot of code from `cvm::atom_group` to `colvar::orientation`. But I think before changing code layout of the orientation CVC and the atom group, we have to allow referencing a CVC in another CVCs.

HanatoK commented:
> > > At this point, there is only a moderate amount of code dedicated to the optimal alignment in the colvaratoms class. And the orientation CVC is already mostly equivalent to what you describe. So maybe this goal could be achieved by adapting the orientation CVC?
> > 
> > 
> > I am not sure. The orientation CVC only gives a quaternion, but the `cvm::atom_group` calculates a rotated frame. If we want to adapt the orientation CVC, then we may have to move a lot of code from `cvm::atom_group` to `colvar::orientation`. But I think before changing code layout of the orientation CVC and the atom group, we have to allow referencing a CVC in another CVCs.
> 
> That's exactly right! Just stated looking at this new PR yesterday, but I wanted to reply on this point, which is that it would require refactoring of the CVC code that we haven't done yet. I think it's okay if this PR focuses on the basic properties like centers of mass for now.

This PR implements almost all operations in the atom group class in CUDA, including the calculations of COM and COG, optimal alignment, rotation, and the rotation derivative. If you would like to restructure the CVC code and move the optimal alignment out of the atom group class, I can adapt the GPU code path accordingly.

HanatoK committed:
opt: avoid unnecessary memsets

HanatoK committed:
fix: properly synchronize the data from GPU to CPU when debug gradients

HanatoK committed:
fix: ensure the fitting group positions are read in case of debug

HanatoK committed:
fix: fix the apply_colvar_force GPU code path

HanatoK committed:
fix: fix the copy of total forces to CPU

HanatoK committed:
WIP

HanatoK committed:
Merge the recent changes from master and fix the derivatives

HanatoK committed:
fix: ensure that the GPU buffer for ref_pos is allocated

HanatoK commented:
There is a race condition and I am still debugging, so I mark this PR as draft.

HanatoK committed:
fix: add the missing dependencies to avoid race condition

HanatoK committed:
fix: try fixing the race condition again

HanatoK committed:
fix: fix the dependency again

HanatoK committed:
fix: read the GPU buffers and calculate the CVC again after debug
gradients

HanatoK committed:
build: add tests

HanatoK committed:
test: update the CUDA stub interface

Backport the fix from colvarproxy_stub and update the proxy description.

HanatoK committed:
build: remove stdc++exp in the CMakeLists.txt

HanatoK committed:
opt: optimize the jacobi_4x4 kernel

- Use template parameters for apply_jacobi and apply_jacobi
- Use absolute value for the diagonal sum
- Sort the eigenvectors partially by only moving the leading eigenvector
  to the first column
- Compute c^2, s^2 and cs in parallel

HanatoK committed:
First (failed) attempt to make jacobo_4x4 to work with float32

HanatoK committed:
debug: add debug code for dependency handling in case of cvm::debug()

HanatoK committed:
refactor: refactor the debug interface and fix the number of atoms of
fitting group

HanatoK committed:
Add the missing colvaratoms_gpu.cpp in the previous commit

HanatoK committed:
perf: use the new approach to project the force to quaternion in GPU
kernel

HanatoK committed:
refactor: remove the anonymous namespace in colvar_gpu_calc.cpp

The functions are reorgranized into the colvarmodule_gpu_calc class.

HanatoK committed:
refactor: move the GPU buffers and data structures to a new class

HanatoK committed:
build: update the build files

HanatoK committed:
fix: fix compute-sanitizer initcheck error

Ensure that the applied forces array is cleared in the stub proxy.

HanatoK committed:
refactor: remove small memsets and memcpys

This commit removes the memsets before COM and COG calculations that
might cause race conditions, and uses temporary buffers for summing COM
and COG. This commit also merges the memcpys that are used to save the
original COG of the main and the fitting groups into the calculations of
COGs.

HanatoK committed:
chore: disable the block reduction if not necessary

HanatoK committed:
refactor: use the new function prepare_dependencies for preparing the
dependencies between CUDA graph nodes instead of the C-style macros

HanatoK committed:
perf: remove unnecessary dependency when computing the COM/COG

HanatoK committed:
refactor: rename the variables and nodes to make them more consistent

HanatoK committed:
doc: update the doxygen of colvaratoms_gpu.h

HanatoK committed:
doc: enable doxygen for COLVARS_CUDA

HanatoK committed:
doc: update doxygen documentation for colvar_gpu_calc

HanatoK committed:
doc: add doxygen documentation for supporting functions in colvars_gpu

HanatoK committed:
doc: document the colvarproxy_gpu class

HanatoK commented:
> The [tests/functional_gpu/run_colvars_test_cuda.cpp](https://github.com/Colvars/colvars/compare/master...HanatoK:colvars_gpu?expand=1#diff-6ff575eee2834fb3ad0e9e130a1398f8feed5bc78bcd8672b8d3ececcd4a0e9b) can be compiled with `-DCOLVARS_DEBUG` and outputs the CUDA graphs. You can run
> 
> ```
> ./run_colvars_test_cuda  ../../input_files/orientationangle-fitgroup_harmonic-fixed/test.in ../../input_files/trajectory.xyz xxx
> ```
> 
> with enabling fit gradients (commenting out `enableFitGradients no`) and get the graphs as follows:
> 
> Read data: [xxx_read_data.pdf](https://github.com/user-attachments/files/21758212/xxx_read_data.pdf)
> 
> Fit gradients: [xxx_fit_gradients.pdf](https://github.com/user-attachments/files/21758214/xxx_fit_gradients.pdf)
> 
> Apply forces: [xxx_apply_forces.pdf](https://github.com/user-attachments/files/21758217/xxx_apply_forces.pdf)

The code PR has been significantly revised to merge many small memset and memcpy nodes into kernels. An updated CUDA graph of reading data is attached:
[xxx_read_data11.pdf](https://github.com/user-attachments/files/22144579/xxx_read_data11.pdf)

HanatoK committed:
test: add a flag to output the forces from proxy

HanatoK committed:
test: save the test output files to functional_gpu

HanatoK committed:
feat: implement GPU kernels for cvm::atom_group

HanatoK committed:
refactor: use the new algorithm to project the forces from C to fitting
group

HanatoK committed:
fix: prepare the rotation derivative for vector CVs

HanatoK committed:
test: use CLI11 for run_colvars_test_cuda

HanatoK committed:
test: test run_colvars_test_cuda with output forces

HanatoK committed:
fix: add the missing GPU kernel for applying forces on fitting groups

HanatoK committed:
fix: remove the debug print

HanatoK commented:
> The failed tests of the GPU code are:
> 
> ```
>           6 - coordnum-aniso_harmonic-fixed_spiff (Failed)
>          10 - coordnum-pairlist_harmonic-fixed_spiff (Failed)
>          12 - coordnum_harmonic-fixed_spiff (Failed)
>          13 - customfunction_harmonic-fixed (Subprocess aborted)
>          14 - customfunction_harmonic-fixed_spiff (Failed)
>         120 - distanceinv_harmonic-fixed_spiff (Failed)
>         138 - distancez-axis-fitgroup_harmonic-fixed_spiff (Failed)
>         146 - eigenvector-difference_harmonic-fixed_spiff (Failed)
>         148 - eigenvector-normalized_harmonic-fixed_spiff (Failed)
>         150 - eigenvector_harmonic-fixed_spiff (Failed)
>         182 - polar_harmonic-fixed_spiff (Failed)
>         186 - rmsd-fitgroup_harmonic-fixed_spiff (Failed)
> ```
> 
> I am still looking into the issues.

Commit https://github.com/Colvars/colvars/pull/826/commits/3703bb1841baa20ee8c2b9f734110a0bc9ead210 with also https://github.com/Colvars/colvars/pull/868 should fix most of the problems except:
```
 13 - customfunction_harmonic-fixed (Subprocess aborted)
 14 - customfunction_harmonic-fixed_spiff (Failed)
```
The custom function does not work because I have not figured out how to add lepton to cmake yet.

HanatoK committed:
test: add lepton to run_colvars_test_cuda

HanatoK committed:
chore: remove the debug line in colvaratoms_gpu.cpp

jhenin commented on src/colvar_gpu_support.cpp:
```diff
@@ -0,0 +1,121 @@
+      error += "\nBacktrace:\n" + std::to_string(std::stacktrace::current()) + "\n";
+#endif
+#endif
+    return cvm::error(error, COLVARS_ERROR);
```
When this detects an error at the time of `colvarproxy` construction, before `colvarmodule` is allocated, `cvm::error()` triggers a segfault, and no error message is printed. Possible solutions:

- make this function avoid using cvm methods
- make the proxy constructor use a safe alternative

jhenin reviewed (commented):
Looking good in general!

How does this interact with scripting? I expected it to be incompatible, but there are changes to `colvarscript`.

jhenin commented:
I see!

Aside from the comment I made, I see no significant obstacles to merging.

HanatoK committed:
fix: use std::cerr in gpuAssert

cvm::error may not be available in the construction of colvarproxy, so I
have to use std::cerr to avoid potential segfaults.

jhenin committed:
fix: make cvm::log & cvm::error safe at all times

Just checking that the static proxy pointer is initialized suffices.
This reverts commit 88166009824633bd636408fac0d792894fe34e59.

jhenin commented:
I've pushed a fix that's more satisfactory to me: make log() and error() check before they try to dereference the proxy pointer! Hopefully this avoids dangerous behaviour in other situations.

HanatoK committed:
fix: fix the typo in the macro

giacomofiorin reviewed (approved):
Looks awesome! Approved following conversation offline.
